### PR TITLE
convert failed pod recreation and rollback without restarts DS e2e tests

### PR DESCRIPTION
**Pull request by crimsonfaith91**
**What this PR does / why we need it**:
This PR converts failed pod recreation and rollback without unnecessary restarts daemonset e2e tests to integration tests.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
xref #TBA

**Release note**:

```release-note
NONE
```

### CODE REVIEW

File: test/integration/daemonset/daemonset_test.go [hunk 1]
```diff
@@ -393,3 +397,267 @@ func TestInsufficientCapacityNodeDaemonDoesNotLaunchPod(t *testing.T) {
 		close(stopCh)
 	}
 }
+
+func getPods(t *testing.T, podClient corev1typed.PodInterface, labelMap map[string]string) *v1.PodList {
```
Comment (crimsonfaith91) [inline test/integration/daemonset/daemonset_test.go]: Should we put all helper functions in a new file `test/integration/daemonset/util.go` like deployment does, or in a more generic new test util file `test/utils/daemonset.go`? I prefer latter because there are some functions that we definitely can share among both e2e and integration tests. However, it seems that latter requires non-tedious code refactoring to existing daemonset e2e tests (it may be wise to do the refactoring in another PR).

File: test/integration/daemonset/daemonset_test.go [hunk 2]
```diff
@@ -393,3 +397,267 @@ func TestInsufficientCapacityNodeDaemonDoesNotLaunchPod(t *testing.T) {
...
+	// Create 2 nodes
+	addNodes(nodeClient, 0, 2, nil, t)
+
+	// Validate everything works correctly, include marking pods
+	// as ready and ensure they launch on every node
+	validateDaemonSetPodsAndMarkReady(podClient, podInformer, 2, t)
+	validateDaemonSetStatus(dsClient, ds.Name, ds.Namespace, 2, t)
+
+	// Manually update image of first pod to the new image
+	// Note: daemonset image update does not propagate to its pods
```
Comment (crimsonfaith91) [inline test/integration/daemonset/daemonset_test.go]: I spent around 2 days finding a way for daemonset image update to propagate to its pods without manually changing the pods, but failed. I even tried to delete the daemonset, but the daemonset failed to get deleted (you can still get the daemonset after the delete operation succeeds). This problem does not exist for deployment and replicaset, so I speculate that I miss out something important that is unique to daemonset.

As this test is about rolling update daemonset, the pods should be automatically deleted and replaced with new ones when there is a change to daemonset spec. However, coming out with something below does not work:
```
// the image change does not propagate to the pods
newDS := updateDS(t, dsClient, dsName, func(update *v1beta1.DaemonSet){
	update.Spec.Template.Spec.Containers[0].Image = newImage
})

// this checks whether there is at least one pod with newImage, but the check fails (all pods still have old image)
checkAtLeastOnePodWithImage(podClient, newImage, t)
```

The pods are not deleted and replaced even after a long polling timeout (3 minutes).

**Comment by crimsonfaith91** [comment]
There is another edge case e2e test named that verifies a rolling update daemonset should adopt existing pods when it is created regardless of `templateGeneration` value. It is very long (especially adding the many required helper functions), so it may be wise to have it in another PR.

I encountered a weird problem that didn't happen when I worked on deployment and replicaset tests - I was not able to manually delete a rolling update daemonset:
```
// set up delete options accordingly
deletePropagationOrphanPolicy := metav1.DeletePropagationOrphan
deleteOptions := &metav1.DeleteOptions{
	PropagationPolicy: &deletePropagationOrphanPolicy,
	Preconditions: metav1.NewUIDPreconditions(string(ds.UID)),
}
// delete operation succeeds with no error
if err = dsClient.Delete(ds.Name, deleteOptions); err != nil {
	t.Fatalf("failed deleting daemonset %q: %v", ds.Name, err)
}
// the daemonset still exists, so this check fails
checkDaemonSetDeleted(dsClient, ds.Name, t)
```

My speculation is that due to nature of daemonset, the controller will get restarted immediately / blocked from being deleted immediately. I discovered that the pods have `BlockOwnerDeletion` field set to true for `controllerRef`, so I manually changed the field to false, but the daemonset still cannot be deleted.

**Comment by crimsonfaith91** [comment]
/assign @janetkuo 
I left 3 comments (they are my questions). Perhaps you can have a look at it before reviewing the codes.

File: test/integration/daemonset/daemonset_test.go (hunk 1 above)
Comment (foxish) [reply test/integration/daemonset/daemonset_test.go]: I had https://github.com/kubernetes/kubernetes/pull/55089/files which put all the workload controllers into an `apps/` directory and had `apps/util.go` for these methods. Let me see if I can revive that.

Comment (crimsonfaith91) [reply test/integration/daemonset/daemonset_test.go]: Great, that would be helpful!

File: test/integration/daemonset/daemonset_test.go [hunk 3]
```diff
@@ -393,3 +397,267 @@ func TestInsufficientCapacityNodeDaemonDoesNotLaunchPod(t *testing.T) {
...
+			pod.Status.Phase = v1.PodFailed
+		})
+
+		// Validate everything works correctly again
+		validateDaemonSetPodsAndMarkReady(podClient, podInformer, 1, t)
+		validateDaemonSetStatus(dsClient, ds.Name, ds.Namespace, 1, t)
+
+		// Validate the failed pod is re-created
+		pods = getPods(t, podClient, testLabels())
+		pod = &pods.Items[0]
```
Comment (foxish) [inline test/integration/daemonset/daemonset_test.go]: I think we can make this check stronger, to ensure that the new replacement pod is on the same node.

File: test/integration/daemonset/daemonset_test.go [hunk 4]
```diff
@@ -393,3 +397,267 @@ func TestInsufficientCapacityNodeDaemonDoesNotLaunchPod(t *testing.T) {
...
+		validateDaemonSetStatus(dsClient, ds.Name, ds.Namespace, 1, t)
+
+		// Set the pod's phase to "Failed"
+		pods := getPods(t, podClient, testLabels())
+		pod := updatePodStatus(t, podClient, pods.Items[0].Name, func(pod *v1.Pod) {
+			pod.Status.Phase = v1.PodFailed
+		})
+
+		// Validate everything works correctly again
+		validateDaemonSetPodsAndMarkReady(podClient, podInformer, 1, t)
```
Comment (foxish) [inline test/integration/daemonset/daemonset_test.go]: isn't marking them as ready going to revert the change we made to pod phase above?

Comment (crimsonfaith91) [reply test/integration/daemonset/daemonset_test.go]: As sync by person, `validateDaemonSetPodsAndMarkReady` manually sets the pod status to be running and ready again. The correct way is to let the daemonset controller to create a new running pod.

File: test/integration/daemonset/daemonset_test.go (hunk 3 above)
Comment (crimsonfaith91) [reply test/integration/daemonset/daemonset_test.go]: Good point! This test only creates one node, so i think the check whether the pod is on the same node is not necessary. However, for the second test which creates 2 nodes, it is necessary to check whether the pod is on the correct node.

**Comment by crimsonfaith91** [comment]
@foxish The newest commit is ready for review!
